Support pgvector as vector store.

ProGM opened the pull request:
I've tried to implement #332 (aka support for pgvector as a vector store).

It's largely based on the Milvus' one.

It uses by default a `vector_l2_ops` index with `lists = 100` and `probes = 10`.
Not sure if it's the optimal setup for GPTCache.

I've implemented the same tests and added a PostgreSQL service to Github Actions using the [official pgvector docker image](https://hub.docker.com/r/ankane/pgvector), so it can be tested on CI.

My knowledge of both system is not super deep, so I'm super-open to feedback and suggestions on how to improve this.

Thanks!

SimFG commented:
@ProGM thanks, some test cases has failed, please check it

SimFG commented on gptcache/manager/vector_data/pgvector.py:
```diff
@@ -0,0 +1,173 @@
+    def bind_processor(self, dialect):
+        def process(value):
+            return value
+        return process
```
maybe a lambda function is better

SimFG commented on gptcache/manager/vector_data/pgvector.py:
```diff
@@ -0,0 +1,173 @@
+        vector store table
+        """
+
+        __tablename__ = table_prefix + "_vector_store"
```
`_pg_vector_store` is better, because there will be other vector store using the `sqlalchemy`

SimFG commented on gptcache/manager/vector_data/pgvector.py:
```diff
@@ -0,0 +1,173 @@
+    def delete(self, ids = None):
+        with self._session() as session:
+            if ids is None:
+                self._query(session).delete()
```
the `ids` is never `None`, and the operation is dangerous. So the `if` can be deleted

ProGM commented:
Hi @SimFG ! Thank you for your review ;)
I've fixed what you asked and rebased dev.

I also changed the Github Actions configuration and this should hopefully fix the tests, but you may need to approve the workflow to make them run, I think.

ProGM commented:
Hi @SimFG, sorry, the pipeline was failing again. Now I fixed it, and tested on my fork.

You can see the output here:
https://github.com/ProGM/GPTCache/actions/runs/4966879957/jobs/8888573740?pr=1

The tests I've added are now passing, but there's still one failing because of Langchain. Not sure if it's related with my change, but I don't think so. Here's the output:

```
FAILED unit_tests/adapter/test_langchain_models.py::test_langchain_chats - TypeError: Can't instantiate abstract class LangChainChat with abstract methods _llm_type
```

SimFG commented:
@ProGM i will check it

SimFG commented:
@ProGM I have fixed it, you can rebase the latest `dev` branch

ProGM commented:
Yay!

Rohianon commented:
Hey, 

I see the latest pgvector support is at 0.4.4, 
![image](https://github.com/zilliztech/GPTCache/assets/41194018/8a539ab8-9fad-4989-b4e9-cb3a3f37bf3a)
Could there be, by any chances that this could be updated from 0.4.1 to match it?

Rohianon commented:
@SimFG I was just wondering why the vector extension you mentioned for stackhero postgres is versioned at 0.4.1 instead of the latest pgvector version of 0.4.4
